cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -214,98 +193,223 @@ public PipelinePlan plan(PipelineSpec spec) {
 
         // if dag1 has any sinks that are a source in dag2, add a connection between the dags
         if (Sets.intersection(dag1.getSinks(), dag2.getSources()).size() > 0) {
-          Set<String> conditionAsSink = Sets.intersection(dag1.getSinks(), conditionBranches.keySet());
-          if (!conditionAsSink.isEmpty()) {
-            // dag1 ends with condition
-            // here we need to add new phase corresponding to the destination condition
-            // also we need to add phase connections with either true or false from dag1 to dag2
-            if (conditionAsSink.size() != 1) {
-              // sink should only have a single stage if its condition
-              throw new IllegalStateException(String.format("Dag '%s' to '%s' has condition as well " +
-                                                              "as non-condition stages in sink which is not " +
-                                                              "allowed.", dag1.getSources(), dag1.getSinks()));
+
+          // check if common node is control node
+          Set<String> controlAsSink = Sets.intersection(dag1.getSinks(), controlNodes);
+          if (!controlAsSink.isEmpty()) {
```

Comment:
it's pretty hard to understand what's happening with so many if elses at so many levels and needing to scroll up to see what the else is matched to. It's easier to read if we do:
```
if (controlAsSink.isEmpty()) {
  phaseConnections.add(new Connection(dag1Name, dag2Name));
  continue;
}
```